File: solutions/python/04/dsu/main.py
class DSU:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def unite(self, a: int, b: int) -> bool:
        a, b = self.find(a), self.find(b)
        if a == b:
            return False
        if self.size[a] < self.size[b]:
            a, b = b, a
        self.parent[b] = a
        self.size[a] += self.size[b]
        return True

    def same(self, a: int, b: int) -> bool:
        return self.find(a) == self.find(b)


def main() -> None:
    dsu = DSU(5)
    dsu.unite(0, 1)
    dsu.unite(2, 3)
    print(dsu.same(0, 1), dsu.same(0, 2))


if __name__ == "__main__":
    main()

File: solutions/go/04/dsu/main.go
package main

import "fmt"

type DSU struct {
	parent []int
	size   []int
}

func NewDSU(n int) *DSU {
	parent := make([]int, n)
	size := make([]int, n)
	for i := range parent {
		parent[i] = i
		size[i] = 1
	}
	return &DSU{parent, size}
}

func (d *DSU) Find(x int) int {
	for d.parent[x] != x {
		d.parent[x] = d.parent[d.parent[x]]
		x = d.parent[x]
	}
	return x
}

func (d *DSU) Unite(a, b int) bool {
	a, b = d.Find(a), d.Find(b)
	if a == b {
		return false
	}
	if d.size[a] < d.size[b] {
		a, b = b, a
	}
	d.parent[b] = a
	d.size[a] += d.size[b]
	return true
}

func (d *DSU) Same(a, b int) bool {
	return d.Find(a) == d.Find(b)
}

func main() {
	d := NewDSU(5)
	d.Unite(0, 1)
	d.Unite(2, 3)
	fmt.Println(d.Same(0, 1), d.Same(0, 2))
}
